// models/categories.js
var mongoose = require('mongoose');

//category schema definition
var categoriesSchema = new mongoose.Schema({
	id: {type: Number, index: {unique: true}},
	category: String
});


var Category = mongoose.model('Category', categoriesSchema);


// Add new category
exports.create = function(model, requestBody, response) {

	// convert from json data structure to model instance
	var category = toCategory(requestBody, model);
	var id = requestBody.id;
	category.save(function(error) {
		if (!error) {
			category.save();
			response.send('Category Added successfully');
		} else {
			console.log(error);
		}
	})
};

// Show all data in category collection
exports.list = function(model, response) {
	model.find({}, function(error, result) {
		if (error) {
			console.error(error);
			return null;
		}
		if (response != null) {
			// response.setHeader('content-type', 'application/json');
			response.end(JSON.stringify(result))
		}
		return JSON.stringify(result);
	});
}

// Find Category based on argument supplied
exports.findByArg = function(model, arg, val, response) {
	model.findOne({
		arg : val
	}, function(error, result) {
		if (error) {
			console.error(error);
			response.writeHead(500, {
				'Content-Type' : 'text/plain'
			});
			response.end('Internal server error');
			return;
		} else {
			if (!result) {
				if (response != null) {
					response.writeHead(404, {
						'Content-Type' : 'text/plain'
					});
					response.end('Not Found');
				}
				return;
			}
			if (response != null) {
				response.setHeader('Content-Type', 'application/json');
				response.send(result);
			}
			console.log(result);
		}
	});
};

// Find Category By Name
exports.findByName = function(model, requestBody, response) {
	model.find({
		category : requestBody.categoryname
		
	}, function(error, result) {
		if (error) {
			console.error(error);
			response.writeHead(500, {
				'Content-Type' : 'text/plain'
			});
			response.end('Internal server error');
			return;
		} else {
			if (!result) {
				if (response != null) {
					response.writeHead(404, {
						'Content-Type' : 'text/plain'
					});
					response.end('Not Found');
				}
				return;
			}
			if (response != null) {
				response.setHeader('Content-Type', 'application/json');
				response.send(result);
			}
			console.log(result);
		}
	});
}

// Remove category based on Id given
exports.remove = function(model, arg, response) {

	model.findOne({
		id : arg
	}, function(error, result) {
		if (!error) {
			result.remove();
			response.send('Category Removed successfully');
		} else {
			console.log(error);
		}
	})
};

// convert from json data structure to model instance
function toCategory(body, Recording) {
	var id_gen = Math.random() * (10000 - 1) + 1;

	return new Recording({
		id : id_gen,
		category : body.category,

	});
}

exports.Category = Category;
