// models/continents.js
var mongoose = require('mongoose');

//continent schema definition
var continentsSchema = new mongoose.Schema({
	id: {type: Number, index: {unique: true}},
	continent: String
});


var Continent = mongoose.model('Continent', continentsSchema);


//Add new continent
exports.create = function(model, requestBody, response) {

	// convert from JSON data structure to model instance
	var continent = toContinent(requestBody, model);
	var id = requestBody.id;
	continent.save(function(error) {
		if (!error) {
			continent.save();
			response.send('Continent Added successfully');
		} else {
			console.log(error);
		}
	})
};

//Show all data in continent collection
exports.list = function(model, response) {
	model.find({}, function(error, result) {
		if (error) {
			console.error(error);
			return null;
		}
		if (response != null) {
			// response.setHeader('content-type', 'application/json');
			response.end(JSON.stringify(result))
		}
		return JSON.stringify(result);
	});
}

//Find Continent based on argument supplied
exports.findByArg = function(model, arg, val, response) {
	model.findOne({
		arg : val
	}, function(error, result) {
		if (error) {
			console.error(error);
			response.writeHead(500, {
				'Content-Type' : 'text/plain'
			});
			response.end('Internal server error');
			return;
		} else {
			if (!result) {
				if (response != null) {
					response.writeHead(404, {
						'Content-Type' : 'text/plain'
					});
					response.end('Not Found');
				}
				return;
			}
			if (response != null) {
				response.setHeader('Content-Type', 'application/json');
				response.send(result);
			}
			console.log(result);
		}
	});
};

//Find Continent By Name
exports.findByName = function(model, requestBody, response) {
	model.find({
		continent : requestBody.continent
		
	}, function(error, result) {
		if (error) {
			console.error(error);
			response.writeHead(500, {
				'Content-Type' : 'text/plain'
			});
			response.end('Internal server error');
			return;
		} else {
			if (!result) {
				if (response != null) {
					response.writeHead(404, {
						'Content-Type' : 'text/plain'
					});
					response.end('Not Found');
				}
				return;
			}
			if (response != null) {
				response.setHeader('Content-Type', 'application/json');
				response.send(result);
			}
			console.log(result);
		}
	});
}

//Remove continent based on Id given
exports.remove = function(model, arg, response) {

	model.findOne({
		id : arg
	}, function(error, result) {
		if (!error) {
			result.remove();
			response.send('Continent Removed successfully');
		} else {
			console.log(error);
		}
	})
};

//convert from JSON data structure to model instance
function toContinent(body, Recording) {
	var id_gen = Math.random() * (10000 - 1) + 1;

	return new Recording({
		id : id_gen,
		continent : body.continent,

	});
};

exports.Continent = Continent;
