// models/memories.js
var mongoose = require('mongoose')
//mongoose.connect('mongodb://localhost');

var recordingsSchema = new mongoose.Schema({
	id : {
		type : Number,
		index : {
			unique : true
		}
	},
	file_address : String,
	caption : String,
	category : [ String ],
	type_id : Number,
	location_id : Number,
	Usersid : Number,
	user_galleries_id : Number,
	created_at : Date,
	upDated_at : Date
});

var Recording = mongoose.model('Recording', recordingsSchema);

exports.create = function(model, requestBody, response){
	//convert from json data structure to model instance
	var memory = toMemory(requestBody, model);
	var id = requestBody.id;
	memory.save(function(error){
		if(!error){
			memory.save();
			response.send('Memory Added successfully');
		}else{
			console.log(error);
		}
	})
};

//Show all data in memories collection
exports.list = function(model, response){
	model.find({}, function(error, result){
		if(error){
			console.error(error);
			return null;
		}if(response!=null){
			//response.setHeader('content-type', 'application/json');
			response.end(JSON.stringify(result))
		}
		return JSON.stringify(result);
	});
}

//convert from json data structure to model instance
function toMemory(body, Recording){
	return new Recording({
		id : body.id,
		file_address : body.file_address,
		caption : body.caption,
		category :body.category,
		type_id : body.type_id,
		location_id : body.location_id,
		Usersid : body.Usersid,
		user_galleries_id : body.user_galleries_id ,
		created_at : body.created_at,
		upDated_at : body.upDated_at
	});
}

exports.Recording = Recording;
